Realm/Tests/RealmTests.mm

```diff
@@ -1423,21 +1423,111 @@ - (void)testIsEmpty {
+    [[NSFileManager defaultManager] removeItemAtPath:fileURL.path error:nil];
+    assert(![[NSFileManager defaultManager] fileExistsAtPath:fileURL.path]);
     [[NSFileManager defaultManager] createFileAtPath:fileURL.path contents:fileContents attributes:nil];
 
     NSError *error;
     RLMRealmConfiguration *configuration = [RLMRealmConfiguration defaultConfiguration];
     configuration.fileURL = fileURL;
     XCTAssertNil([RLMRealm realmWithConfiguration:configuration error:&error], @"Invalid database");
     XCTAssertNotNil(error, @"Should populate error object");
+    // Validate error contents
+    XCTAssertEqualObjects(error.domain, RLMErrorDomain);
+    XCTAssertEqual(error.code, RLMErrorFileAccess);
+    XCTAssert([error.userInfo[@"Underlying"] rangeOfString:@"Realm file has bad size"].location != NSNotFound);
+}
+
+- (void)testRealmFileAccessBadEncryptionKey
+{
+    // Create keys
+    char goodKeyArray[64];
+    char badKeyArray[64];
+    memset(goodKeyArray, 'a', 64);
+    memset(badKeyArray, 'z', 64);
+
+    NSData *goodKey = [NSData dataWithBytes:goodKeyArray length:64];
+    NSData *badKey = [NSData dataWithBytes:badKeyArray length:64];
+
+    NSError *error = nil;
+
+    // Set up file URLs and purge old state
+    NSFileManager *manager = [NSFileManager defaultManager];
+    NSURL *goodURL = [NSURL fileURLWithPath:RLMRealmPathForFile(@"filename.realm")];
+    NSURL *badURL = [[goodURL URLByDeletingLastPathComponent] URLByAppendingPathComponent:@"badfile.realm"];
+    [manager removeItemAtPath:goodURL.path error:nil];
+    [manager removeItemAtPath:badURL.path error:nil];
+    assert(![manager fileExistsAtPath:goodURL.path] && ![manager fileExistsAtPath:badURL.path]);
+
+    // Create the first Realm using the good encryption key
+    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
+    config.encryptionKey = goodKey;
+    config.fileURL = goodURL;
+    RLMRealm *goodRealm = [RLMRealm realmWithConfiguration:config error:&error];
+    XCTAssertNotNil(goodRealm);
+    XCTAssertNil(error);
+
+    // Copy the Realm file on disk so we can attempt to open a second Realm
+    [manager copyItemAtURL:goodURL toURL:badURL error:nil];
+    assert([manager fileExistsAtPath:badURL.path]);
+
+    // Attempt opening the second Realm
+    error = nil;
+    config = [RLMRealmConfiguration defaultConfiguration];
+    config.encryptionKey = badKey;
+    config.fileURL = badURL;
+    RLMRealm *badRealm = [RLMRealm realmWithConfiguration:config error:&error];
+
+    XCTAssertNil(badRealm);
+    XCTAssertNotNil(error);
+    // Validate error contents
+    XCTAssertEqualObjects(error.domain, RLMErrorDomain);
+    XCTAssertEqual(error.code, RLMErrorFileAccess);
+    XCTAssert([error.userInfo[@"Underlying"] rangeOfString:@"Realm file decryption failed"].location != NSNotFound);
+}
+
+- (void)testRealmFileAccessFileIsDirectory
+{
+    NSURL *fileURL = [NSURL fileURLWithPath:RLMRealmPathForFile(@"filename.realm")];
```

Comment:
This test could be simplified by creating the directory at `RLMTestRealmURL()`, as RLMTestCase takes care of deleting anything that already exists at that path and also cleans it up after the test.